### PR TITLE
Streaming support for sync and reconciliation

### PR DESCRIPTION
**Pull request by CassOnMars**
## Why is this change needed?

Add support for streaming interfaces for sync and reconciliation. Shuttle change to immediately follow.

## Merge Checklist

_Choose all relevant options below by adding an `x` now or at any time before submitting for review_

- [x] PR title adheres to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard
- [x] PR has a [changeset](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#35-adding-changesets)
- [x] PR has been tagged with a change label(s) (i.e. documentation, feature, bugfix, or chore)
- [ ] PR includes [documentation](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#32-writing-docs) if necessary.

---

## PR-Codex overview
This PR updates dependencies, adds streaming support for sync/reconciliation, and introduces a new `useStreaming` option in the Hubble application.

### Detailed summary
- Added `ClientDuplexStream` in `hub-nodejs`
- Updated versions in `core` and `hub-nodejs`
- Added streaming methods in `rpc.proto`
- Introduced `useStreaming` option in `hubble` application

> The following files were skipped due to too many changes: `packages/hub-web/src/generated/rpc.ts`, `packages/hub-nodejs/src/generated/rpc.ts`, `apps/hubble/src/network/sync/syncEnginePerf.test.ts`, `apps/hubble/src/network/utils/networkConfig.test.ts`, `apps/hubble/src/network/sync/syncEngine.ts`, `apps/hubble/src/rpc/server.ts`, `packages/hub-web/src/generated/request_response.ts`, `packages/hub-nodejs/src/generated/request_response.ts`, `packages/core/src/protobufs/generated/request_response.ts`, `apps/hubble/src/network/sync/multiPeerSyncEngine.test.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: apps/hubble/src/network/sync/syncEngine.ts [hunk 1]
```diff
@@ -152,6 +164,332 @@ type SyncStatus = {
...
+  usingRPC = false;
+
+  rpcClient: HubRpcClient;
+  stream?: ClientDuplexStream<StreamSyncRequest, StreamSyncResponse>;
+
+  constructor(rpcClient: HubRpcClient) {
+    this.rpcClient = rpcClient;
+    this.rpcClient.streamSync().then((result) => {
+      if (result.isErr()) {
+        log.warn("encountered error when attempting to establish sync for stream, failing over to RPC", result.error);
```
Comment (sanjayprabhu) [inline apps/hubble/src/network/sync/syncEngine.ts:177]: I think we should also allow overriding using the network config so we can globally toggle stream sync on and off. 

We can remove this later once we know it's stable. Would be good to have the feature flag for the initial rollout.

File: apps/hubble/src/network/sync/multiPeerSyncEngine.test.ts [hunk 2]
```diff
@@ -322,65 +335,98 @@ describe("Multi peer sync engine", () => {
...
+    console.log("36");
     expect(await syncEngine2.trie.rootHash()).toEqual(beforeRootHash);
   });
 
   test("audit should fail if peer withholds messages", async () => {
+    console.log("b1");
     await expect(engine1.mergeOnChainEvent(custodyEvent)).resolves.toBeDefined();
+    console.log("b2");
     await expect(engine1.mergeOnChainEvent(signerEvent)).resolves.toBeDefined();
+    console.log("b3");
```
Comment (sanjayprabhu) [inline apps/hubble/src/network/sync/multiPeerSyncEngine.test.ts:415]: remove all the console logs?

File: apps/hubble/src/network/sync/multiPeerSyncEngine.test.ts [hunk 3]
```diff
@@ -142,7 +142,7 @@ describe("Multi peer sync engine", () => {
 
     server1 = new Server(hub1, engine1, syncEngine1);
     port1 = await server1.start();
-    clientForServer1 = getInsecureHubRpcClient(`127.0.0.1:${port1}`);
+    clientForServer1 = new FailoverStreamSyncClient(getInsecureHubRpcClient(`127.0.0.1:${port1}`));
```
Comment (sanjayprabhu) [inline apps/hubble/src/network/sync/multiPeerSyncEngine.test.ts:145]: For this test file, can we have the tests run in both stream mode and rpc mode to make sure both codepaths are working as expected?

File: apps/hubble/src/rpc/server.ts [hunk 4]
```diff
@@ -1517,6 +1651,324 @@ export default class Server {
           }
         }
       },
+      streamSync: async (stream: ServerDuplexStream<StreamSyncRequest, StreamSyncResponse>) => {
+        const timeout = setTimeout(async () => {
```
Comment (sanjayprabhu) [inline apps/hubble/src/rpc/server.ts:1655]: Should we also have a limit on the number of concurrent streams?

**Review by sanjayprabhu — COMMENTED** [review]
High level looks good, left some questions. Will do one more pass tomorrow.

File: apps/hubble/src/network/sync/multiPeerSyncEngine.test.ts (hunk 2 above)
Comment (CassOnMars) [reply apps/hubble/src/network/sync/multiPeerSyncEngine.test.ts:415]: whoops, thought i killed those off, been cross eyed trying to tackle the last issue 😅

File: apps/hubble/src/rpc/server.ts (hunk 4 above)
Comment (CassOnMars) [reply apps/hubble/src/rpc/server.ts:1655]: We don't presently have a limit on sync-related rpcs, so this remains in parity on resources, possibly better given there's lower potential for NLB dropping weirdness

File: apps/hubble/src/network/sync/syncEngine.ts (hunk 1 above)
Comment (CassOnMars) [reply apps/hubble/src/network/sync/syncEngine.ts]: added!

**Pull request merged**